lib/src/main/scala/spinal/lib/Stream.scala

```diff
@@ -1085,101 +1090,277 @@ object StreamFifo{
   def apply[T <: Data](dataType: T, depth: Int) = new StreamFifo(dataType,depth)
 }
 
-class StreamFifo[T <: Data](val dataType: HardType[T], val depth: Int) extends Component {
+/**
+  * Improved StreamFifo over the original implementation
+  * - latency of 0, 1, 2 cycles
+  * - Better timings
+  *
+  * @param dataType
+  * @param depth Number of element stored in the fifo, Note that if withAsyncRead==false, then one extra transaction can be stored
+  * @param withAsyncRead Read the memory using asyncronous read port (ex distributed ram). If false, add 1 cycle latency
+  * @param withBypass Bypass the push port to the pop port when the fifo is empty. If false, add 1 cycle latency
+  */
+class StreamFifo[T <: Data](val dataType: HardType[T],
+                            val depth: Int,
+                            val withAsyncRead : Boolean = false,
+                            val withBypass : Boolean = false,
+                            val allowExtraMsb : Boolean = true,
+                            val forFMax : Boolean = false,
+                            val useVec : Boolean = false) extends Component {
   require(depth >= 0)
-  val io = new Bundle {
+  // bypass not supported for sync read case
+  if(!withAsyncRead) require(!withBypass)
```

Comment:
should we add a note like "use of withAsyncRead=false requires withBypass=false as well"?

Reply:
Yes. Same for `useVec`, as it requires` withAsyncRead`.
We should add this to the code:
```scala
if (useVec) require (withAsyncRead)
```

Reply:
Yes, on it.